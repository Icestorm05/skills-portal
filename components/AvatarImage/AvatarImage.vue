<template lang="pug">
    img(v-if="src"
        :src="src"
        alt="Avatar").AvatarImage
</template>

<script lang="ts">
import Component from "nuxt-class-component";
import objectFitImages from "object-fit-images";
import Vue from "vue";
import { Prop, Watch } from "vue-property-decorator";

@Component
export default class AvatarImage extends Vue {
  @Prop({ type: String, default: "" })
  private defaultImagePrefix: string;
  @Prop({ type: String, default: "765-default-avatar" })
  private defaultImageName: string;
  @Prop({ type: String, default: "jpg" })
  private defaultImageExtension: string;
  @Prop({ type: Boolean, default: false })
  private renderDefaultImageAsBase64: boolean;

  @Prop({ type: String, default: "avatars" })
  private prefix: string;
  @Prop({ type: String })
  private imageName: string | undefined;
  @Prop({ type: String, default: "jpg" })
  private extension: string;
  @Prop({ type: Boolean, default: false })
  private renderAsBase64: boolean;

  private src = "";

  @Watch("defaultImageName")
  @Watch("imageName", { immediate: true })
  private async imageNameChanged() {
    try {
      if (!this.imageName) {
        throw new Error("No image specified.");
      }
      this.src = this.renderAsBase64
        ? await this.base64(this.image, this.extension)
        : this.image;
    } catch (err) {
      if (!this.defaultImageName) {
        throw new Error("No default image specified.");
      }
      this.src = this.renderDefaultImageAsBase64
        ? await this.base64(
            this.defaultImage,
            this.defaultImageExtension
          ).catch()
        : this.defaultImage;
    } finally {
      objectFitImages();
    }
  }

  private async base64(image: string, extension: string) {
    const options = { baseURL: "", responseType: "arraybuffer" };
    const buffer = await (this as any).$axios.$get(image, options);
    const base64 = btoa(
      new Uint8Array(buffer).reduce(
        (data, byte) => data + String.fromCharCode(byte),
        ""
      )
    );
    const mimetype = `image/${extension}`;
    return `data:${mimetype};base64,${base64}`;
  }

  private get defaultImage() {
    const prefix = this.defaultImagePrefix ? `/${this.defaultImagePrefix}` : "";
    return `${prefix}/${this.defaultImageName}.${this.defaultImageExtension}`;
  }

  private get image() {
    const prefix = this.prefix ? `/${this.prefix}` : "";
    return `${prefix}/${this.imageName}.${this.extension}`;
  }
}
</script>

<style lang="stylus" scoped>
.AvatarImage {
  object-fit: cover;
  object-position: top;
}
</style>
