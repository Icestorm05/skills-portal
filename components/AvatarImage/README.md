# Avatar Image

## Functionality

The avatar image component attempts to display an image, with a fallback if failed.

### Props

`<avatar-image>` supports the following custom component attributes:

| attribute | type | description
| --- | --- | ---
| `defaultImagePrefix` | String *optional* | The URL to prefix to the default image.
| `defaultImageName` | String *optional* | The name of the fallback image. Defaults to `765-default-avatar`.
| `defaultImageExtension` | String *optional* | The extension of the fallback image. Defaults to `jpg`.
| `renderDefaultImageAsBase64` | Boolean *optional* | Whether to render the fallback image as base64 or not. Defaults to `false`.

| `prefix` | String *optional* | The URL to prefix to the image. Defaults to `avatars`.
| `imageName` | String *optional* | The name of the image. If no image is specified, it will default to the fallback image.
| `extension` | String *optional* | The extension of the image. Defaults to `jpg`.
| `renderAsBase64` | Boolean *optional* | Whether to render the image as base64 or not. Defaults to `true`.

### Example

```html
<!-- This will try to render an img tag as base64 using avatarImages/John_Doe.gif.
     If it errors, it will instead render an img tag as base64 using images/fallback.png. -->
<avatar-image defaultImagePrefix="images"
              defaultImageName="fallback"
              defaultImageExtension="png"
              renderDefaultImageAsBase64
              prefix="avatarImages"
              imageName="John_Doe"
              extension="gif"
              renderAsBase64/>
```

```ts
import Vue from 'vue';
import Component from 'nuxt-class-component';
import AvatarImage from '~/components/AvatarImage/AvatarImage.vue';
@Component({
    components: {
        AvatarImage,
    },
})
export default class GenericComponent extends Vue {}
```