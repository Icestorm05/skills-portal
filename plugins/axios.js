import axios from "axios";
import Cookies from "js-cookie";

export default async ({ $axios, store, redirect, req, route }, inject) => {
  let cancelToken = axios.CancelToken.source();
  const cancelAllRequests = () => {
    cancelToken.cancel();
    cancelToken = axios.CancelToken.source();
    inject("cancelToken", cancelToken);
  };
  inject("cancelAllRequests", cancelAllRequests);
  inject("cancelToken", cancelToken);

  const token = process.server ? req.cookies.token : Cookies.get("token");
  if (token) {
    $axios.setToken(token, "Bearer");
  }

  $axios.onError(err => {
    const code = parseInt(err.response && err.response.status, 10);
    if (code === 401) {
      store.commit("auth/resetToken");
      store.commit("profile/reset");
      store.commit("snackbar/show", {
        message: "Session timed out. Please re-login.",
        color: "error"
      });
      Cookies.remove("token");
      $axios.setToken(false);
      const query = route.query.redirect;
      redirect(`/login?redirect=${query ? query : route.path}`);
    }
  });
};
