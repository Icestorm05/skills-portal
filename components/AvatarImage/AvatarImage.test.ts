import AvatarImage from "./AvatarImage.vue";
import axios from "axios";

import Vue from "vue";
import { shallow } from "@vue/test-utils";

function $get() {
  return new ArrayBuffer(1);
}

declare const expect: jest.Expect;

describe("Avatar Image", () => {
  test("is a Vue instance", () => {
    const wrapper = shallow(AvatarImage);
    expect(wrapper.isVueInstance()).toBeTruthy();
  });

  test("default snapshot has not changed", async () => {
    const wrapper = shallow(AvatarImage);
    await wrapper.vm.$nextTick();
    expect(wrapper.html()).toMatchSnapshot();
  });

  test("image snapshot has not changed", async () => {
    const wrapper = shallow(AvatarImage, {
      propsData: { imageName: "test" },
      mocks: { $axios: { $get } }
    });
    await wrapper.vm.$nextTick();
    expect(wrapper.html()).toMatchSnapshot();
  });

  test("default base64 snapshot has not changed", async () => {
    const wrapper = shallow(AvatarImage, {
      propsData: { renderDefaultImageAsBase64: true },
      mocks: { $axios: { $get } }
    });
    await wrapper.vm.$nextTick();
    await wrapper.vm.$nextTick();
    expect(wrapper.html()).toMatchSnapshot();
  });

  test("image base64 snapshot has not changed", async () => {
    const wrapper = shallow(AvatarImage, {
      propsData: { renderAsBase64: true, imageName: "test" },
      mocks: { $axios: { $get } }
    });
    await wrapper.vm.$nextTick();
    await wrapper.vm.$nextTick();
    expect(wrapper.html()).toMatchSnapshot();
  });

  test("passing no image name prop renders the default avatar", async () => {
    const wrapper = shallow(AvatarImage, { mocks: { $axios: { $get } } });
    await wrapper.vm.$nextTick();
    expect(wrapper.find(".AvatarImage").attributes().src).toEqual(
      "/765-default-avatar.jpg"
    );
  });

  test("passing image name prop renders the image as base64", async () => {
    const wrapper = shallow(AvatarImage, {
      propsData: { imageName: "test", renderAsBase64: true },
      mocks: { $axios: { $get } }
    });
    await wrapper.vm.$nextTick();
    await wrapper.vm.$nextTick();
    expect(wrapper.find(".AvatarImage").attributes().src).toEqual(
      "data:image/jpg;base64,AA=="
    );
  });

  test("passing default image name changes the default avatar", async () => {
    const wrapper = shallow(AvatarImage, {
      propsData: { defaultImageName: "test" },
      mocks: { $axios: { $get } }
    });
    await wrapper.vm.$nextTick();
    expect(wrapper.find(".AvatarImage").attributes().src).toEqual("/test.jpg");
  });

  test("passing extension changes the extension of the image", async () => {
    const wrapper = shallow(AvatarImage, {
      propsData: { imageName: "test", extension: "png", renderAsBase64: true },
      mocks: { $axios: { $get } }
    });
    await wrapper.vm.$nextTick();
    await wrapper.vm.$nextTick();
    expect(wrapper.find(".AvatarImage").attributes().src).toEqual(
      "data:image/png;base64,AA=="
    );
  });

  test("passing extension does not change the extension of the default avatar", async () => {
    const wrapper = shallow(AvatarImage, {
      propsData: { extension: "gif" },
      mocks: { $axios: { $get } }
    });
    await wrapper.vm.$nextTick();
    expect(wrapper.find(".AvatarImage").attributes().src).toEqual(
      "/765-default-avatar.jpg"
    );
  });

  test("passing default image extension changes the extension of the default avatar", async () => {
    const wrapper = shallow(AvatarImage, {
      propsData: { defaultImageExtension: "gif" },
      mocks: { $axios: { $get } }
    });
    await wrapper.vm.$nextTick();
    expect(wrapper.find(".AvatarImage").attributes().src).toEqual(
      "/765-default-avatar.gif"
    );
  });

  test("computed default image uses default image prefix, default image name and default image extension", async () => {
    const wrapper = shallow(AvatarImage, {
      propsData: {
        defaultImagePrefix: "test",
        defaultImageName: "test",
        defaultImageExtension: "gif"
      },
      mocks: { $axios: { $get } }
    });
    await wrapper.vm.$nextTick();
    expect((wrapper.vm as any).defaultImage).toEqual("/test/test.gif");
  });

  test("computed image uses prefix, image name and extension", async () => {
    const wrapper = shallow(AvatarImage, {
      propsData: { imageName: "test", extension: "png", prefix: "avi" },
      mocks: { $axios: { $get } }
    });
    await wrapper.vm.$nextTick();
    expect((wrapper.vm as any).image).toEqual("/avi/test.png");
  });

  test("imageNameChanged fn is fired on mount", () => {
    const spy = jest.fn();
    const wrapper = shallow(AvatarImage, {
      methods: { imageNameChanged: spy },
      mocks: { $axios: { $get } }
    });
    expect(spy).toHaveBeenCalled();
  });

  test("imageNameChanged fn is fired on image name prop change", () => {
    const spy = jest.fn();
    const wrapper = shallow(AvatarImage, {
      methods: { imageNameChanged: spy },
      mocks: { $axios: { $get } }
    });
    expect(spy).toHaveBeenCalledTimes(1);
    wrapper.setProps({ imageName: "test" });
    expect(spy).toHaveBeenCalledTimes(2);
  });

  test("no img tag is rendered immediately when image name prop is passed in with renderAsBase64", async () => {
    const wrapper = shallow(AvatarImage, {
      propsData: { imageName: "test", renderAsBase64: true },
      mocks: { $axios: { $get } }
    });
    expect(wrapper.find(".AvatarImage").exists()).toBeFalsy();
    await wrapper.vm.$nextTick();
    await wrapper.vm.$nextTick();
    expect(wrapper.find(".AvatarImage").exists()).toBeTruthy();
  });

  test("passing in renderDefaultImageAsBase64 renders the default image as base64", async () => {
    const wrapper = shallow(AvatarImage, {
      propsData: { renderDefaultImageAsBase64: true },
      mocks: { $axios: { $get } }
    });
    await wrapper.vm.$nextTick();
    await wrapper.vm.$nextTick();
    expect(wrapper.find(".AvatarImage").attributes().src).toEqual(
      "data:image/jpg;base64,AA=="
    );
  });

  test("not passing in renderAsBase64 renders the image as specified url", async () => {
    const wrapper = shallow(AvatarImage, {
      propsData: { imageName: "test" },
      mocks: { $axios: { $get } }
    });
    await wrapper.vm.$nextTick();
    await wrapper.vm.$nextTick();
    expect(wrapper.find(".AvatarImage").attributes().src).toEqual(
      "/avatars/test.jpg"
    );
  });

  test("passing in no default image or image displays nothing", async () => {
    const wrapper = shallow(AvatarImage, {
      propsData: { defaultImageName: "" },
      mocks: { $axios: { $get } }
    });
    await wrapper.vm.$nextTick();
    await wrapper.vm.$nextTick();
    expect(wrapper.find(".AvatarImage").exists()).toBeFalsy();
  });

  test("image with no prefix renders correctly", async () => {
    const wrapper = shallow(AvatarImage, {
      propsData: { imageName: "test", prefix: "" },
      mocks: { $axios: { $get } }
    });
    await wrapper.vm.$nextTick();
    await wrapper.vm.$nextTick();
    expect(wrapper.find(".AvatarImage").attributes().src).toEqual("/test.jpg");
  });
});
