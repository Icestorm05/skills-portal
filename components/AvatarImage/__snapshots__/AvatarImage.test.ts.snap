// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar Image default base64 snapshot has not changed 1`] = `<img src="data:image/jpg;base64,AA==" alt="Avatar" class="AvatarImage">`;

exports[`Avatar Image default snapshot has not changed 1`] = `<img src="/765-default-avatar.jpg" alt="Avatar" class="AvatarImage">`;

exports[`Avatar Image image base64 snapshot has not changed 1`] = `<img src="data:image/jpg;base64,AA==" alt="Avatar" class="AvatarImage">`;

exports[`Avatar Image image snapshot has not changed 1`] = `<img src="/avatars/test.jpg" alt="Avatar" class="AvatarImage">`;
